### PR TITLE
Begin upgrading tab shortcuts

### PR DESCRIPTION
**Pull request by chunkybanana**
So, what I've done here is made it so that sequences of letters/numbers/underscores can get replaced with a corresponding character. For example, typing `sum` then pressing tab becomes `∑`.

This still needs a whole lot more work - it's just quite time-consuming to write out aliases. I think it'll be nice to have though.

If y'all want to help, imitate the existing shortcuts. Make sure to allow underscores separating words.

Marking this as draft for now.

### CODE REVIEW

File: static/main.js [hunk 1]
```diff
@@ -194,6 +194,63 @@ const aliases = {
...
+    "°": ["deg","complex"],
+    "•": ["log",/r(epeat)?_?cha?r/,"mold",/c(aps)?_?tr(ans(fer)?)?/],
+    "ß": [/cond(ition(al)?)?(_?execute)?/],
+    "†": [/call(_?func(tion)?)?/,/len_?p(rime)?_?f(actors)?/,/py(thon)?_?exec|python/,/vec(torised)?_?not/],
+    "€": [/split_?on/,/fill_?(by_?)?coord(inate)?s/],
+    "½": ["half","halve"],
+    // ∆ø and other digraphs skipped
+    "↔": [/fix(ed_?point)?/,/inter(sect(ion)?)?/,/comb(ination|o)s?/],
+    "¢": [/inf(inite)?_?repl(ace(ment)?)?/,/apply_?(at_?)ind(ice)?s/],
+    "∆¢": ["carmicheal"],
```
Comment (Steffan153) [inline static/main.js:230]: carmicheal lol
```suggestion
    "∆¢": ["carmichael"],
```

File: static/main.js [hunk 2]
```diff
@@ -476,14 +533,21 @@ function initCodeMirror() {
...
+                }
                 const k = lines[cur.line].slice(cur.ch - 2, cur.ch);
                 const t = Object.entries(aliases).find(x => x[1].includes(k));
                 if (t) {
-                    const l = [...lines[cur.line]];
-                    l.splice(cur.ch - 2, 2, t[0]);
-                    lines[cur.line] = l.join('');
-                    cm.setValue(lines.join("\n"));
-                    cm.setCursor({ ...cur, ch: cur.ch - 1 });
+                    // Sorry Steffan, you've been usurped by the robots.
```
Comment (Steffan153) [inline static/main.js:549]: I was using Copilot...

File: static/main.js [hunk 3]
```diff
@@ -194,6 +194,63 @@ const aliases = {
     "‟": ["''"]
 };
 
+// While the two byte aliases attempt to identify what characters look like,
```
Comment (Steffan153) [inline static/main.js:197]: Btw, some of the two-byte aliases are also what they do, for example logical OR is `or`.

File: static/main.js [hunk 4]
```diff
@@ -194,6 +194,63 @@ const aliases = {
...
+    "⌐": [/compl([ei]ment)?/,/comma(_?split)?/],
+    "æ": [/(is_?)?prime/,/case(_?of)?/],
+    "ʀ": [/inc(lusive)?_?(z(ero)?|0)_?r(ange)?/,/is_?alpha/],
+    "ʁ": [/exc(lusive)?_?(z(ero)?|0)_?r(ange)?/],
+    "ɾ": [/inc(lusive)?_?(o(ne)?|1)_?r(ange)?/,/vec_?upper/],
+    "ɽ": [/exc(lusive)?_?(o(ne)?|1)_?r(ange)?/,/vec_?lower/],
+    "øɽ": [/(just(ify)?|align)_?r(ight)?|r(ight)?_?(just(ify)?|align)/],
+    "ƈ": [/cho(ice|ose)|bin(omial)?_?coeff(icient)?/,/rand(om)?_?cho(ice|ose)/,/set_?(eq(ual)?|same)/,/drop_?w(hile)?/],
+    "∆ƈ": ["npr",/n_?p(ick)?_?r/],
+    "∞": [/pal(indrom(e|ise))?/],
```
Comment (Steffan153) [inline static/main.js:240]: Y'all Australians
```suggestion
    "∞": [/pal(indrom(e|i[sz]e))?/],
```

File: static/main.js [hunk 5]
```diff
@@ -194,6 +194,63 @@ const aliases = {
...
+    "∧": [/(logic_?)?and/],
+    "ø∧": [/g?canvas/],
+    "⟑": [/apply(_?map)?/,/force_?eval/],
+    "∨": [/(logic_?)?or/],
+    "k∨": [/a(ll)?_?vowels?/],
+    "Þ∨": [/m(ulti(set)?)?_?diff/],
+    "⟇": [/remove_?index/,/rm_?index/,"rmi"],
+    "k⟇": ["codepage"],
+    "ø⟇": [/(codepage|cp)_?(get|find|help|util)/],
+    "÷": [/(i(tem)?_?)split/],
```
Comment (Steffan153) [inline static/main.js:214]:
```suggestion
    "÷": [/(i(tem)?_?)split/, /p(ush)?_?ea(ch)?/],
```

File: static/main.js (hunk 2 above)
Comment (chunkybanana) [reply static/main.js:549]: Lol

File: static/main.js (hunk 5 above)
Comment (chunkybanana) [reply static/main.js:214]: pea lol

**Review by Lyxal — APPROVED** [review]

File: static/main.js [hunk 6]
```diff
@@ -194,6 +194,273 @@ const aliases = {
...
+    "∆ṁ": ["median"],
+    "øṁ": [/vert(ical)?_?mirror/],
+    "Þṁ": [/mold_?(no|without)_?rep(eat)?/],
+    "ṅ": [/join_?(nil|nothing)/,/insig(nificant)?/,/pad_?bytes?/,/bytes?_?pad/,/f(irst)?_?num/],
+    "kṅ": ["8192"],
+    "ȯ": [/slice(_?end)?/,/f(irst)?_?n_?nums?/,/vert(ical)?_?merge/],
+    "ṗ": [/(p(ower)?|sub)sets?/],
+    "kṗ": [/brac(ket)?_?pairs?/],
+    "∆ṗ": [/prev(ious)?_?prime/],
+    // how does shortcut øṗ?
```
Comment (Steffan153) [inline static/main.js:292]:
```suggestion
    "øṗ": [/pal(indromi[sz]e)?_?cen(ter)?_?new(lines?)?/],
```

**Review by Steffan153 — APPROVED** [review]

File: static/main.js [hunk 7]
```diff
@@ -194,6 +194,273 @@ const aliases = {
...
+    "Þ⇩": [/(is_?)?strict(ly)?_?(dec(reasing)?|desc(ending)?)/],
+    "Ǎ": [/(remove|rm)_?non_?alpha/,"2pow"],
+    "ǎ": [/substr(ing)?s?/,/n(th)?_?prime/],
+    "Ǐ": [/dist(inct)?_?prime_?fact(or)?s?/,/append_?(first|head)/],
+    "ǐ": [/prime_?fact(or)?s?/,/sent(ence)?_?case/],
+    "∆ǐ" : [/prime_?exp(onents?)?/],
+    "Ǒ": [/multiplic(ity)?/,/(remove|rm)_?fix(ed)?(_?point)?/,/find_?tru(e|thy)/],
+    "ǒ": [/mod_?(3|three)|(3|three)_?mod/,/chunks?_?2|2_?chunks?/],
+    "Ǔ": [/rot(ate)?_?r(ight)?/],
+    "ÞǓ": [/conn(ected)?_?uniq(ue|ify)?/],
```
Comment (Steffan153) [inline static/main.js:445]:
```suggestion
    "ÞǓ": [/conn(ected)?_?uniqu(e|ify)?/],
```

uniqify lol

**Review by Steffan153 — CHANGES_REQUESTED** [review]
we should also add some for the digraphs that use typable chars.

File: static/main.js [hunk 8]
```diff
@@ -406,10 +406,10 @@ const other_aliases = {
...
     "↲": [/shift_?l(eft)?/,/l(eft)?_?pad/,/pad_?l(eft)?/],
     "ø↲": [/c(ustom)?_?pad_?r(ight)?/],
-    "⋏": [/bit(wise)?_?and/,/pad_?c(enter)?|c(enter)?_?pad/],
-    "⋎": [/bit(wise)?_?or/,/(rm|remove)_?ind(ex)?/,"merge"],
-    "꘍": [/bitwise_?xor/,/l(even(shtein)?)?_?dist(ance)?/,/(add|prepend|append)_?spaces?/],
-    "ꜝ": [/bitwise_?not/,/filter_?fals[ye]/,/any_?upper(case)?/],
+    "⋏": [/bit(wi[sz]e)?_?and/,/pad_?c(enter)?|c(enter)?_?pad/],
+    "⋎": [/bit(wi[sz]e)?_?or/,/(rm|remove)_?ind(ex)?/,"merge"],
+    "꘍": [/bitwi[sz]e_?xor/,/l(even(shtein)?)?_?dist(ance)?/,/(add|prepend|append)_?spaces?/],
+    "ꜝ": [/bitwi[sz]e_?not/,/filter_?fals[ye]/,/any_?upper(case)?/],
```
Comment (Steffan153) [inline static/main.js:412]: What on earth, bitwize is not a word.

File: static/main.js [hunk 9]
```diff
@@ -194,6 +194,372 @@ const aliases = {
...
+    "øM": [/vert(ical)_?flip_?palindrom(is)?e/],
+    "ÞM": [/max(imal)?_?ind(ice)?s?/],
+    "ÞN": [/alt(ernating)?_?neg(ation)?/],
+    "∆P": [/solve_?poly/],
+    "øP": [/plurali[sz]e/],
+    "∆Q": [/solve_?quad(ratic)?/],
+    "∆R": [/to_?rad(ians)?/],
+    "øR": [/strip_?right/],
+    // How does ÞR
+    "∆S": ["arcsin"],
```
Comment (Steffan153) [inline static/main.js:507]:
```suggestion
    "∆S": [/arc_?sine?/],
```

File: static/main.js [hunk 10]
```diff
@@ -194,6 +194,372 @@ const aliases = {
...
+    "¨=": [/eq(ual)?_?under/],
+    "Þ<": [/less_inc(reasing)?/],
+    "¨<": [/strict_?(l(ess)?_?t(han)?)/],
+    "¨>": [/strict_?(g(reater)?_?t(han)?)?/],
+    "¨?": [/read_?stdin/],
+    "øA": [/letter_?num(ber)?/,/num(ber)?_?letter/],
+    "ÞA": [/adj(acency)?_?mat(rix)?/],
+    "øB": ["bracketify"],
+    "ÞB": [/rand(om)?_?bits/],
+    "∆C": ["arccos"],
```
Comment (Steffan153) [inline static/main.js:480]:
```suggestion
    "∆C": [/arc_?cos(ine)?/],
```

File: static/main.js [hunk 11]
```diff
@@ -194,6 +194,372 @@ const aliases = {
...
+    "∆o": [/n(th)?_?ord(inal)?/],
+    "Þo": [/(all_?)ord(inal)?s/],
+    "øp": [/starts?_?with/],
+    "Þp": [/(all_?)primes/],
+    "¨p": [/over_?pairs/],
+    "ør": [/strip_?right_?by/],
+    "Þr": [/(remove|rm)_?(last|tail)_?p(repend)?_?(0|zero)/],
+    "∆s": ["sin"],
+    "øs": [/starts_?(with_?)?set/],
+    "∆t": ["tan"],
```
Comment (Steffan153) [inline static/main.js:556]:
```suggestion
    "∆t": [/tan(gent)?/],
```

File: static/main.js [hunk 12]
```diff
@@ -194,6 +194,372 @@ const aliases = {
...
+    "Þe": [/mat(rix)?_?exp(onent(iation)?)?/],
+    "∆f": [/n(th)?_?fib(onacci)?/],
+    "øf": [/ends?_?(with_?)?set/],
+    "Þf": [/flat_?by/],
+    "Þg": ["shortest"],
+    "∆i": [/n(th)?_?pi_?digit/],
+    "Þi": [/m(ulti)?dim(ension)?_?ind(ex)?/],
+    "¨i": [/if_?else/],
+    "Þj": ["depth"],
+    "∆l": ["log2"],
```
Comment (Steffan153) [inline static/main.js:542]:
```suggestion
    "∆l": [/log_?2/],
```

File: static/main.js [hunk 13]
```diff
@@ -194,6 +194,372 @@ const aliases = {
...
+    "∆P": [/solve_?poly/],
+    "øP": [/plurali[sz]e/],
+    "∆Q": [/solve_?quad(ratic)?/],
+    "∆R": [/to_?rad(ians)?/],
+    "øR": [/strip_?right/],
+    // How does ÞR
+    "∆S": ["arcsin"],
+    "øS": [/strip_?(left|right)/],
+    "ÞS": ["sublists"],
+    "∆T": ["arctan"],
```
Comment (Steffan153) [inline static/main.js:510]:
```suggestion
    "∆T": [/arc_?tan(gent)?/],
```

File: static/main.js [hunk 14]
```diff
@@ -194,6 +194,372 @@ const aliases = {
...
+    "ÞM": [/max(imal)?_?ind(ice)?s?/],
+    "ÞN": [/alt(ernating)?_?neg(ation)?/],
+    "∆P": [/solve_?poly/],
+    "øP": [/plurali[sz]e/],
+    "∆Q": [/solve_?quad(ratic)?/],
+    "∆R": [/to_?rad(ians)?/],
+    "øR": [/strip_?right/],
+    // How does ÞR
+    "∆S": ["arcsin"],
+    "øS": [/strip_?(left|right)/],
```
Comment (Steffan153) [inline static/main.js:508]:
```suggestion
    "øS": [/strip_?both/],
```
it's not stripleft or stripright, it's both

File: static/main.js [hunk 15]
```diff
@@ -194,6 +194,372 @@ const aliases = {
...
+    "k∪": [/l(ow(ercase)?)?_?vowels?_?y/],
+    "Þ∪": [/m(ulti)?(set)?_?union/],
+    "∩": ["trans","tp","transpose"],
+    "k∩": [/vowels?_?y/],
+    "Þ∩": [/m(ulti)?(set)?_?intersection/],
+    "⊍": [/set_?(diff(erence)?|xor)/],
+    "k⊍": [/up(per(case)?)?_?vowels?_?y/],
+    "Þ⊍": [/m(ulti)?(set)?_?(diff(erence)?|xor)/],
+    "£": [/s(et)?_?reg(ister)?/],
+    "¨£": ["starmap"],
```
Comment (Steffan153) [inline static/main.js:431]:
```suggestion
    "¨£": [/star_?map/, /zip_?with/],
```

**Review by Steffan153 — CHANGES_REQUESTED** [review]

File: static/main.js (hunk 8 above)
Comment (chunkybanana) [reply static/main.js:412]: Oh. I'm bad at American :P

File: static/main.js (hunk 14 above)
Comment (chunkybanana) [reply static/main.js:508]: Damnit, copilot

File: static/main.js (hunk 8 above)
Comment (cgccuser) [reply static/main.js:412]: Keep it. This is exactly how we Americanz zpeak

File: static/main.js [hunk 16]
```diff
@@ -194,6 +194,372 @@ const aliases = {
...
+    "∆ǐ" : [/prime_?exp(onents?)?/],
+    "Ǒ": [/multiplic(ity)?/,/(remove|rm)_?fix(ed)?(_?point)?/,/find_?tru(e|thy)/],
+    "ǒ": [/mod_?(3|three)|(3|three)_?mod/,/chunks?_?2|2_?chunks?/],
+    "Ǔ": [/rot(ate)?_?r(ight)?/],
+    "ÞǓ": [/conn(ected)?_?uniqu(e|ify)?/],
+    "ǔ": [/rot(ate)?_?l(eft)?/],
+    "Þǔ": ["untruth"],
+    "⁽": [/l(am(bda)?)?_?1|1_?l(am(bda)?)?/],
+    "‡": [/l(am(bda)?)?_?2|2_?l(am(bda)?)?/],
+    "≬": [/l(am(bda)?)?_?3|3_?l(am(bda)?)?/],
```
Comment (cgccuser) [inline static/main.js:450]: Maybe also `n_byte_lambda`?

Comment (chunkybanana) [reply static/main.js:450]: WIll do

File: static/main.js [hunk 17]
```diff
@@ -194,6 +194,372 @@ const aliases = {
...
+    "∆M": ["mode"],
+    "øM": [/vert(ical)_?flip_?palindrom(is)?e/],
+    "ÞM": [/max(imal)?_?ind(ice)?s?/],
+    "ÞN": [/alt(ernating)?_?neg(ation)?/],
+    "∆P": [/solve_?poly/],
+    "øP": [/plurali[sz]e/],
+    "∆Q": [/solve_?quad(ratic)?/],
+    "∆R": [/to_?rad(ians)?/],
+    "øR": [/strip_?right/],
+    // How does ÞR
```
Comment (Steffan153) [inline static/main.js:506]:
```suggestion
    "ÞR": [/(zero|0)_?cum(ulative)?_?sums?/],
```

File: static/main.js [hunk 18]
```diff
@@ -194,6 +194,372 @@ const aliases = {
...
+    "øf": [/ends?_?(with_?)?set/],
+    "Þf": [/flat_?by/],
+    "Þg": ["shortest"],
+    "∆i": [/n(th)?_?pi_?digit/],
+    "Þi": [/m(ulti)?_?dim(ension(al)?)?_?ind(ex)?/],
+    "¨i": [/if_?else/],
+    "Þj": ["depth"],
+    "∆l": [/log_?2/],
+    "øl": [/strip_?left_?by/],
+    // øm?
```
Comment (Steffan153) [inline static/main.js:544]:
```suggestion
    "øm": [/m(ir(ror)?)?_?c(enter)?_?newlines?/],
```

**Review by Steffan153 — APPROVED** [review]

**Review by Steffan153 — CHANGES_REQUESTED** [review]
you forgot `∆p`

**Comment by chunkybanana** [comment]
> you forgot `∆p`

Done

**Review by Steffan153 — APPROVED** [review]

**Pull request merged**